File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.cpp
class Solution {
public:
    int maxSumAfterPartitioning(vector<int>& arr, int k) {
          int n = arr.size();
          int dp[n + 1];
          memset(dp, 0, sizeof(dp));
          for(int i= 1; i<=n; i++){
              int K = k;
              int cur_max = 0;
              int cur_ans = 0;
              for(int K = 1; K <= k && i - K >= 0; K++){
                  cur_max = max(cur_max, arr[i-K]);
                  cur_ans = max(cur_ans, dp[i-K] + K * cur_max);
              }
              dp[i] = cur_ans;
          }
           return dp[n];
    }
};

File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr, k):
        N = len(arr)
        K = k + 1

        dp = [0] * K

        for start in range(N - 1, -1, -1):
            curr_max = 0
            end = min(N, start + k)

            for i in range(start, end):
                curr_max = max(curr_max, arr[i])
                dp[start % K] = max(dp[start % K], dp[(i + 1) % K] + curr_max * (i - start + 1))

        return dp[0]

